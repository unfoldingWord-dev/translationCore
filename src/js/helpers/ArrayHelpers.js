/**
 * @description - Groups consecutive numbers in an array
 * @param {Array} numbers - array of numbers to be grouped
 * @returns {Array} - grouped array of array of consecutive numbers
 */
export const groupConsecutiveNumbers = (numbers) => (
  numbers.reduce(function(accumulator, currentValue, currentIndex, originalArray) {
    if (currentValue) { // ignore undefined entries
      // if this iteration is consecutive to the last, add it to the previous run
      if (currentValue - originalArray[currentIndex - 1] === 1) {
        accumulator[accumulator.length - 1].push(currentValue);
      } else { // the start of a new run including first element
        // create a new subarray with this as the start
        accumulator.push([currentValue]);
      }
    }
    return accumulator; // return state for next iteration
  }, [])
);
/**
 * @description - Deletes indices from an array safely
 * @param {Array} array - array elements to delete from
 * @returns {Array} - the resulting array after indexes were safely removed
 */
export const deleteIndices = (array, indices) => {
  let _array = JSON.parse(JSON.stringify(array));
  indices.sort( (a,b) => b - a );
  indices.forEach(index => {
    if (index >= 0) _array.splice(index, 1);
  });
  return _array;
};
