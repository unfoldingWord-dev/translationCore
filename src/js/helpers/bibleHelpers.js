/* eslint-disable no-console */
import Path from 'path-extra';
import fs from 'fs-extra';
import BooksOfBible from '../../../tcResources/books';
import {getLatestVersionInPath} from "./ResourcesHelpers";

/**
 *
 * @param {string} bookAbbr - The book abbreviation to convert
 */
export function convertToFullBookName(bookAbbr) {
  if (!bookAbbr) return;
  return BooksOfBible[bookAbbr.toString().toLowerCase()];
}

/**
 *
 * @param {string} projectBook - Book abbreviation
 */
export function isOldTestament(projectBook) {
  var passedBook = false;
  for (var book in BooksOfBible) {
    if (book === projectBook) passedBook = true;
    if (BooksOfBible[book] === 'Malachi' && passedBook) {
      return true;
    }
  }
  return false;
}

/**
 * determine
 * @param bookId
 * @return {{resourceLanguage: string, bibleID: string}}
 */
export function getOLforBook(bookId) {
  const isOT = isOldTestament(bookId);
  const languageId = (isOT) ? 'he' : 'grc';
  const bibleId = (isOT) ? 'uhb' : 'ugnt';
  return {languageId, bibleId};
}

/**
 * This checks if a project is missing verses.
 * @param projectDir
 * @param bookId
 * @param resourceDir
 * @return {boolean}
 */
export const isProjectMissingVerses = (projectDir, bookId, resourceDir) => {
  try {
    let languageId = 'en';
    const resourcePath = Path.join(resourceDir, languageId, 'bibles', 'ult');
    const versionPath = getLatestVersionInPath(resourcePath) || resourcePath;
    const indexLocation = Path.join(versionPath, 'index.json');
    const expectedVerses = fs.readJSONSync(indexLocation);
    const actualVersesObject = {};
    const currentFolderChapters = fs.readdirSync(Path.join(projectDir, bookId));
    let chapterLength = 0;
    for (let currentChapterFile of currentFolderChapters) {
      let currentChapter = Path.parse(currentChapterFile).name;
      if (!parseInt(currentChapter)) continue;
      chapterLength++;
      let verseLength = 0;
      try {
        let currentChapterObject = fs.readJSONSync(
          Path.join(projectDir, bookId, currentChapterFile)
        );
        for (let verseIndex in currentChapterObject) {
          let verse = currentChapterObject[verseIndex];
          if (verse && verseIndex > 0) verseLength++;
        }
      } catch (e) {
        console.warn(e);
      }
      actualVersesObject[currentChapter] = verseLength;
    }
    actualVersesObject.chapters = chapterLength;
    let currentExpectedVerese = expectedVerses[bookId];
    return (
      JSON.stringify(currentExpectedVerese) !==
      JSON.stringify(actualVersesObject)
    );
  } catch (e) {
    console.warn(
      'ult index file not found missing verse detection is invalid. Please delete ~/translationCore/resources folder'
    );
    return false;
  }
};
