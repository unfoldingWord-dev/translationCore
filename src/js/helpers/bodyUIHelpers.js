export function getIconColorFromIndex(stepIndex) {
  let colorsArray = [];
  colorsArray.push(stepIndex === 0 ? "black" : stepIndex > 0 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 1 ? "black" : 1 > stepIndex ? "grey" : stepIndex > 1 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 2 ? "black" : 2 > stepIndex ? "grey" : stepIndex > 2 ? "var(--accent-color-dark)" : "");
  colorsArray.push(stepIndex === 3 ? "black" : 3 > stepIndex ? "grey" : stepIndex > 3 ? "var(--accent-color-dark)" : "");
  return colorsArray;
}