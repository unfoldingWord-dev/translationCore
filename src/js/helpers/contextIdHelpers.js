import path from 'path';
import fs from 'fs-extra';
import { getQuoteOccurrencesInVerse } from 'selections';
// constants
const INDEX_DIRECTORY = path.join('.apps', 'translationCore', 'index');

/**
 * @description this action validates if the contextId is valid by checking to see if the quote occurrence exists
 * @param {object} state - current state to get the verse out of
 * @param {object} contextId - the contextId object.
 * @param {String} bibleId - the id/name of the bible to get the chapter/verse text to check for the string occurrences
 * @return {bool} returns if the contextId is valid.
 */
export const validateContextIdQuote = (state, contextId, bibleId) => {
  let valid = false;

  if (contextId && bibleId && contextId.quote) {
    const { chapter, verse } = contextId.reference;
    const { quote, occurrence } = contextId;
    const verseText = state.resourcesReducer.bibles[bibleId][chapter][verse];
    const occurrences = getQuoteOccurrencesInVerse(verseText, quote);
    valid = occurrence <= occurrences;
  }

  if (!contextId.quote) {
    valid = true;
  }
  return valid;
};

/**
 * find path in index for current contextId
 * @param {String} projectSaveLocation
 * @param {String} toolName
 * @param {String} bookId
 * @return {*}
 */
export function getContextIdPathFromIndex(projectSaveLocation, toolName, bookId) {
  return path.join(projectSaveLocation, INDEX_DIRECTORY, toolName, bookId, 'currentContextId', 'contextId.json');
}

/**
 * Writes the context id to the disk.
 * @param {object} state - store state object.
 * @param {object} contextId
 */
export const saveContextId = (state, contextId) => {
  try {
    let { projectSaveLocation } = state.projectDetailsReducer;
    let toolName = contextId ? contextId.tool : undefined;
    let bookId = contextId ? contextId.reference.bookId : undefined;

    if (projectSaveLocation && toolName && bookId) {
      let savePath = getContextIdPathFromIndex(projectSaveLocation, toolName, bookId);
      fs.outputJson(savePath, contextId, { spaces: 2 });
    } else {
      // saveCurrentContextId: missing required data
    }
  } catch (err) {
    console.warn(err);
  }
};

/**
 * make sure context IDs are for same verse.  Optimized over isEqual()
 * @param {Object} contextId1
 * @param {Object} contextId2
 * @return {boolean} returns true if context IDs are for same verse
 */
export function isSameVerse(contextId1, contextId2) {
  return (contextId1.reference.chapter === contextId2.reference.chapter) &&
    (contextId1.reference.verse === contextId2.reference.verse);
}
