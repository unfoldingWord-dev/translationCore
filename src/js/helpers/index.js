import { generateTimestamp } from './TimestampGenerator.js';
import * as loadHelpers from './LoadHelpers.js';

export {
  generateTimestamp,
  loadHelpers,
};
