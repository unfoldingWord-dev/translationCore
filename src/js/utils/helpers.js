import Path from 'path';
import * as fs from 'fs-extra';
import ManifestGenerator from '../components/core/create_project/ProjectManifest';
import BooksOfBible from '../components/core/BooksOfBible';

module.exports.isOldTestament = function (projectBook) {
    var passedBook = false;
    for (var book in BooksOfBible) {
        if (book == projectBook) passedBook = true;
        if (BooksOfBible[book] == "Malachi" && passedBook) {
            return true;
        }
    }
    return false;
}

module.exports.checkIfValidBetaProject = function (manifest) {
    if (manifest && manifest.project) return manifest.project.id == "eph" || manifest.project.id == "tit";
    else if (manifest && manifest.ts_project) return manifest.ts_project.id == "eph" || manifest.ts_project.id == "tit";
}

/**
 * 
 * @param {*} path 
 * @param {*} file 
 */
export function loadFile(path, file) {
    try {
        var manifest = fs.readJsonSync(Path.join(path, file));
        return manifest;
    }
    catch (e) {
        return null
    }
}

/**
 * @description - Generates and saves a translationCore manifest file
 * @param {string} projectSaveLocation - Filepath of where the translationCore manifest file will
 * be saved. Must be an ABSOLUTE PATH
 * @param {object} data - The translationCore manifest data to be saved
 * @param {object} tsManifest - The translationStudio manifest data loaded from a translation
 * studio project
 */
module.exports.saveManifest = function (projectSaveLocation, link, oldManifest, currentUser) {
    var data = {
        user: [currentUser],
        repo: link
    }
    var manifest;
    try {
        var manifestLocation = Path.join(projectSaveLocation, 'manifest.json');
        if (oldManifest.package_version == '3') {
            //some older versions of ts-manifest have to be tweaked to work
            manifest = this.fixManifestVerThree(oldManifest);
        } else {
            manifest = ManifestGenerator(data, oldManifest);
        }
        fs.outputJsonSync(manifestLocation, manifest);
    }
    catch (err) {
        console.error(err);
    }
    return manifest;
}

/**
 * @desription - Uses the tc-standard format for projects to make package_version 3 compatible
 * @param oldManifest - The name of an employee.
 */
module.exports.fixManifestVerThree = function (oldManifest) {
    var newManifest = {};
    try {
        for (var oldElements in oldManifest) {
            newManifest[oldElements] = oldManifest[oldElements];
        }
        newManifest.finished_chunks = oldManifest.finished_frames;
        newManifest.ts_project = {};
        newManifest.ts_project.id = oldManifest.project_id;
        newManifest.ts_project.name = this.convertToFullBookName(oldManifest.project_id);
        for (var el in oldManifest.source_translations) {
            newManifest.source_translations = oldManifest.source_translations[el];
            var parameters = el.split("-");
            newManifest.source_translations.language_id = parameters[1];
            newManifest.source_translations.resource_id = parameters[2];
            break;
        }
    } catch (e) {
        console.error(e);
    }
    return newManifest;
}

module.exports.convertToFullBookName = function (bookAbbr) {
    if (!bookAbbr) return;
    return BooksOfBible[bookAbbr.toString().toLowerCase()];
}