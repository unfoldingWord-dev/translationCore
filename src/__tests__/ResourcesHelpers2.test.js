/* eslint-env jest */
import path from 'path';
import fs from 'fs-extra';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import _ from 'lodash';
// helpers
import { getContextIdPathFromIndex } from '../js/helpers/contextIdHelpers';
import {
  areResourcesNewer,
  copySourceContentUpdaterManifest,
  extractZippedResourceContent,
  findArticleFilePath,
  getAvailableScripturePaneSelections,
  getResourcesNeededByTool,
  loadArticleData,
  updateSourceContentUpdaterManifest,
  updateGroupIndexForGl,
} from '../js/helpers/ResourcesHelpers';
// constants
import {
  APP_VERSION,
  ORIGINAL_LANGUAGE,
  PROJECTS_PATH,
  STATIC_RESOURCES_PATH,
  TARGET_BIBLE,
  TARGET_LANGUAGE,
  TC_VERSION,
  TRANSLATION_ACADEMY,
  TRANSLATION_HELPS,
  TRANSLATION_WORDS,
  TRANSLATION_NOTES,
  USER_RESOURCES_PATH,
} from '../js/common/constants';
jest.mock('fs-extra');
jest.mock('adm-zip');

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('getResourcesNeededByTool()', () => {
  it('should work', () => {
    const bookId = 'gal';
    const store = {
      resourcesReducer: {
        bibles: {},
        translationHelps: {},
        lexicons: {},
      },
      contextIdReducer: {
        contextId: {
          reference: {
            bookId: bookId,
            chapter:1,
          },
        },
      },
      settingsReducer: {
        toolsSettings: {
          ScripturePane: {
            currentPaneSettings: [
              {
                bibleId: TARGET_BIBLE,
                languageId: TARGET_LANGUAGE,
              }, {
                bibleId: 'ugnt',
                languageId: ORIGINAL_LANGUAGE,
              }, {
                bibleId: 'ust',
                languageId: 'en',
              }, {
                bibleId: 'ult',
                languageId: 'en',
              },
            ],
          },
        },
      },
    };
    loadMockFsWithProjectAndResources();

    // when
    const resourceList = getResourcesNeededByTool(store, bookId);

    // then
    expect(resourceList).toMatchSnapshot();
  });
});

describe('getAvailableScripturePaneSelections', () => {
  beforeAll(() => {
    fs.__resetMockFS();
    loadMockFsWithProjectAndResources();
    fs.ensureDirSync(path.join(USER_RESOURCES_PATH, 'en'));
  });

  it('should work', () => {
    const bookId = 'gal';
    const store = mockStore({
      resourcesReducer: {
        bibles: { targetLanguage: { targetBible: { manifest: {} } } },
        translationHelps: {},
        lexicons: {},
      },
      contextIdReducer: {
        contextId: {
          reference: {
            bookId: bookId,
            chapter:1,
          },
        },
      },
      settingsReducer: {
        toolsSettings: {
          ScripturePane: {
            currentPaneSettings: [
              {
                bibleId: TARGET_BIBLE,
                languageId: TARGET_LANGUAGE,
              }, {
                bibleId: 'ugnt',
                languageId: ORIGINAL_LANGUAGE,
              }, {
                bibleId: 'ust',
                languageId: 'en',
              }, {
                bibleId: 'ult',
                languageId: 'en',
              },
            ],
          },
        },
      },
    });
    const resourceList = [];

    // when
    store.dispatch(getAvailableScripturePaneSelections(resourceList));

    // then
    expect(cleanupResources(resourceList)).toMatchSnapshot();
  });

  it('should work OT', () => {
    const bookId = 'jol';
    const store = mockStore({
      resourcesReducer: {
        bibles: { targetLanguage: { targetBible: { manifest: {} } } },
        translationHelps: {},
        lexicons: {},
      },
      contextIdReducer: {
        contextId: {
          reference: {
            bookId: bookId,
            chapter:1,
          },
        },
      },
      settingsReducer: {
        toolsSettings: {
          ScripturePane: {
            currentPaneSettings: [
              {
                bibleId: TARGET_BIBLE,
                languageId: TARGET_LANGUAGE,
              }, {
                bibleId: 'uhb',
                languageId: ORIGINAL_LANGUAGE,
              }, {
                bibleId: 'ust',
                languageId: 'en',
              }, {
                bibleId: 'ult',
                languageId: 'en',
              },
            ],
          },
        },
      },
    });

    const resourceList = [];

    // when
    store.dispatch(getAvailableScripturePaneSelections(resourceList));

    // then
    expect(cleanupResources(resourceList)).toMatchSnapshot();
  });

  it('getAvailableScripturePaneSelections() should work without targetBible loaded into resources', () => {
    const bookId = 'gal';
    const store = mockStore({
      resourcesReducer: {
        bibles: {},
        translationHelps: {},
        lexicons: {},
      },
      contextIdReducer: {
        contextId: {
          reference: {
            bookId: bookId,
            chapter:1,
          },
        },
      },
      settingsReducer: {
        toolsSettings: {
          ScripturePane: {
            currentPaneSettings: [
              {
                bibleId: TARGET_BIBLE,
                languageId: TARGET_LANGUAGE,
              }, {
                bibleId: 'ugnt',
                languageId: ORIGINAL_LANGUAGE,
              }, {
                bibleId: 'ust',
                languageId: 'en',
              }, {
                bibleId: 'ult',
                languageId: 'en',
              },
            ],
          },
        },
      },
    });
    const resourceList = [];

    // when
    store.dispatch(getAvailableScripturePaneSelections(resourceList));

    // then
    expect(cleanupResources(resourceList)).toMatchSnapshot();
  });
});

describe('areResourcesNewer()', () => {
  beforeEach(() => {
    fs.__resetMockFS();
  });

  test('same date should return false', () => {
    // given
    const bundledDate = '2019-04-02T19:10:02.492Z';
    const userDate = '2019-04-02T19:10:02.492Z';
    const expectedNewer = false;
    loadSourceContentUpdaterManifests(bundledDate, userDate);

    // when
    const results = areResourcesNewer();

    // then
    expect(results).toEqual(expectedNewer);
  });

  test('newer bundled date should return true', () => {
    // given
    const bundledDate = '2019-04-02T19:10:02.492Z';
    const userDate = '2018-04-02T19:10:02.492Z';
    const expectedNewer = true;
    loadSourceContentUpdaterManifests(bundledDate, userDate);

    // when
    const results = areResourcesNewer();

    // then
    expect(results).toEqual(expectedNewer);
  });

  test('newer user date should return false', () => {
    // given
    const bundledDate = '2019-04-02T19:10:02.492Z';
    const userDate = '2019-08-02T19:10:02.492Z';
    const expectedNewer = false;
    loadSourceContentUpdaterManifests(bundledDate, userDate);

    // when
    const results = areResourcesNewer();

    // then
    expect(results).toEqual(expectedNewer);
  });

  test('missing user resource manifest should return true', () => {
    // given
    const bundledDate = '2019-04-02T19:10:02.492Z';
    const userDate = null;
    const expectedNewer = true;
    loadSourceContentUpdaterManifests(bundledDate, userDate);

    // when
    const results = areResourcesNewer();

    // then
    expect(results).toEqual(expectedNewer);
  });

  test('missing bundled resource manifest should return false', () => {
    // given
    const bundledDate = null;
    const userDate = '2019-04-02T19:10:02.492Z';
    const expectedNewer = false;
    loadSourceContentUpdaterManifests(bundledDate, userDate);

    // when
    const results = areResourcesNewer();

    // then
    expect(results).toEqual(expectedNewer);
  });

  test('same date, but missing version should return true', () => {
    // given
    const bundledDate = '2019-04-02T19:10:02.492Z';
    const userDate = '2019-04-02T19:10:02.492Z';
    const expectedNewer = true;
    const appVersion = null;
    loadSourceContentUpdaterManifests(bundledDate, userDate, appVersion);

    // when
    const results = areResourcesNewer();

    // then
    expect(results).toEqual(expectedNewer);
  });

  test('same date, but different version should return true', () => {
    // given
    const bundledDate = '2019-04-02T19:10:02.492Z';
    const userDate = '2019-04-02T19:10:02.492Z';
    const expectedNewer = true;
    const appVersion = '1.1.0';
    loadSourceContentUpdaterManifests(bundledDate, userDate, appVersion);

    // when
    const results = areResourcesNewer();

    // then
    expect(results).toEqual(expectedNewer);
  });
});

describe('updateSourceContentUpdaterManifest()', () => {
  beforeEach(() => {
    fs.__resetMockFS();
  });

  test('should update date and app version if manifest missing', () => {
    // given
    const dateStr = '1997-12-17T08:24:00.000Z';
    const manifestPath = path.join(USER_RESOURCES_PATH,
      'source-content-updater-manifest.json');
    expect(fs.existsSync(manifestPath)).not.toBeTruthy();

    // when
    updateSourceContentUpdaterManifest(dateStr);

    // then
    const manifest = fs.readJSONSync(manifestPath);
    expect(manifest[TC_VERSION]).toEqual(APP_VERSION);
    expect(manifest.modified).toEqual(dateStr);
  });

  test('should update date if manifest present', () => {
    // given
    const dateStr = '1997-12-17T08:24:00.000Z';
    const userDate = '2019-04-02T19:10:02.492Z';
    loadSourceContentUpdaterManifests('', userDate);
    const manifestPath = path.join(USER_RESOURCES_PATH,
      'source-content-updater-manifest.json');
    expect(fs.existsSync(manifestPath)).toBeTruthy();

    // when
    updateSourceContentUpdaterManifest(dateStr);

    // then
    const manifest = fs.readJSONSync(manifestPath);
    expect(manifest[TC_VERSION]).toEqual(APP_VERSION);
    expect(manifest.modified).toEqual(dateStr);
  });

  test('should update app version if manifest present', () => {
    // given
    const dateStr = '1997-12-17T08:24:00.000Z';
    const initialAppVersion = '1.0.1';
    loadSourceContentUpdaterManifests('', dateStr, initialAppVersion);
    const manifestPath = path.join(USER_RESOURCES_PATH,
      'source-content-updater-manifest.json');
    expect(fs.existsSync(manifestPath)).toBeTruthy();

    // when
    updateSourceContentUpdaterManifest(dateStr);

    // then
    const manifest = fs.readJSONSync(manifestPath);
    expect(manifest[TC_VERSION]).toEqual(APP_VERSION);
    expect(manifest.modified).toEqual(dateStr);
  });
});

describe('copySourceContentUpdaterManifest()', () => {
  beforeEach(() => {
    fs.__resetMockFS();
  });

  test('should update app version after copy', () => {
    // given
    const dateStr = '1997-12-17T08:24:00.000Z';
    loadSourceContentUpdaterManifests(dateStr, null, null);
    const staticManifestPath = path.join(STATIC_RESOURCES_PATH,
      'source-content-updater-manifest.json');
    expect(fs.existsSync(staticManifestPath)).toBeTruthy();

    // when
    copySourceContentUpdaterManifest(dateStr);

    // then
    const manifestPath = path.join(USER_RESOURCES_PATH,
      'source-content-updater-manifest.json');
    const manifest = fs.readJSONSync(manifestPath);
    expect(manifest[TC_VERSION]).toEqual(APP_VERSION);
    expect(manifest.modified).toEqual(dateStr);
  });
});

describe('extractZippedResourceContent', () => {
  it('works as expected', () => {
    const EN_ULB_PATH = path.join(USER_RESOURCES_PATH, 'en', 'ult');
    const versionPath = path.join(USER_RESOURCES_PATH, 'en', 'ult', 'v11');
    const zippedBooks = path.join(EN_ULB_PATH, 'v11', 'books.zip');
    const isBible = true;

    fs.__setMockFS({
      [EN_ULB_PATH]: ['v11'],
      [versionPath]: [],
      [zippedBooks]: [],
    });

    extractZippedResourceContent(EN_ULB_PATH, isBible);
    expect(fs.existsSync(zippedBooks)).toBeFalsy();
  });
});

describe('findArticleFilePath()', () => {
  it('findArticleFilePath for abel in en', () => {
    loadMockFsWithProjectAndResources();
    const filePath = findArticleFilePath(TRANSLATION_WORDS, 'abel', 'en');
    const expectedPath = path.join(USER_RESOURCES_PATH, 'en', TRANSLATION_HELPS, TRANSLATION_WORDS, 'v10', 'names', 'articles', 'abel.md');
    expect(filePath).toEqual(expectedPath);
  });

  it('findArticleFilePath for a non-existing file', () => {
    loadMockFsWithProjectAndResources();
    const filePath = findArticleFilePath(TRANSLATION_WORDS, 'does-not-exist', 'en');
    expect(filePath).toBeNull();
  });

  it('findArticleFilePath for abraham which is not in Hindi, but search hindi first', () => {
    loadMockFsWithProjectAndResources();
    const filePath = findArticleFilePath(TRANSLATION_WORDS, 'abomination', 'hi');
    const expectedPath = path.join(USER_RESOURCES_PATH, 'hi', TRANSLATION_HELPS, TRANSLATION_WORDS, 'v8.1', 'kt', 'articles', 'abomination.md');
    expect(filePath).toEqual(expectedPath);
  });

  it('findArticleFilePath for abraham which is not in Hindi, but search hindi first', () => {
    loadMockFsWithProjectAndResources();
    const filePath = findArticleFilePath(TRANSLATION_WORDS, 'abraham', 'hi');
    const expectedPath = path.join(USER_RESOURCES_PATH, 'en', TRANSLATION_HELPS, TRANSLATION_WORDS, 'v10', 'names', 'articles', 'abraham.md');
    expect(filePath).toEqual(expectedPath);
  });

  it('findArticleFilePath for tA translate-names which is not in Hindi so should return English', () => {
    loadMockFsWithProjectAndResources();
    const filePath = findArticleFilePath(TRANSLATION_ACADEMY, 'translate-names', 'hi');
    const expectedPath = path.join(USER_RESOURCES_PATH, 'en', TRANSLATION_HELPS, TRANSLATION_ACADEMY, 'v9', 'translate', 'translate-names.md');
    expect(filePath).toEqual(expectedPath);
  });

  it('findArticleFilePath for tW abraham but giving a wrong category should return null', () => {
    loadMockFsWithProjectAndResources();
    const filePath = findArticleFilePath(TRANSLATION_WORDS, 'abraham', 'en', 'kt');
    expect(filePath).toBeNull();
  });
});


describe('loadArticleData()', () => {
  it('loadArticleData for tW abraham giving correct category', () => {
    loadMockFsWithProjectAndResources();
    const articleId = 'abraham';
    const category = 'names';
    const content = loadArticleData(TRANSLATION_WORDS, articleId, 'en', category);
    const notExpectedContent = '# Article Not Found: '+articleId+' #\n\nCould not find article for '+articleId;
    expect(content).toBeTruthy();
    expect(content).not.toEqual(notExpectedContent);
  });

  it('loadArticeData for tW abraham but giving a wrong category should return not found message', () => {
    loadMockFsWithProjectAndResources();
    const articleId = 'abraham';
    const category = 'kt';
    const content = loadArticleData(TRANSLATION_WORDS, articleId, 'en', category);
    const expectedContent = '# Article Not Found: '+articleId+' #\n\nCould not find article for '+articleId;
    expect(content).toEqual(expectedContent);
  });

  it('loadArticleData for tW abraham with no category', () => {
    loadMockFsWithProjectAndResources();
    const articleId = 'abraham';
    const content = loadArticleData(TRANSLATION_WORDS, articleId, 'en');
    const notExpectedContent = '# Article Not Found: '+articleId+' #\n\nCould not find article for '+articleId;
    expect(content).toBeTruthy();
    expect(content).not.toEqual(notExpectedContent);
  });

  it('loadArticleData for tA translate-names with no category and hindi should still find (English) content', () => {
    loadMockFsWithProjectAndResources();
    const articleId = 'translate-names';
    const content = loadArticleData(TRANSLATION_ACADEMY, articleId, 'hi');
    const notExpectedContent = '# Article Not Found: '+articleId+' #\n\nCould not find article for '+articleId;
    expect(content).toBeTruthy();
    expect(content).not.toEqual(notExpectedContent);
  });
});

describe('updateGroupIndexForGl()', () => {
  const bookId = 'gal';
  const manifest_ = {
    'generator':{ 'name':'tc-desktop','build':'' },
    'target_language':{
      'id':'en','name':'English','direction':'ltr',
    },
    'ts_project':{ 'id':bookId,'name':'Galatians' },
    'project':{ 'id':bookId,'name':'Galatians' },
    'type':{ 'id':'text','name':'Text' },
    'time_created':'2018-01-31T19:19:27.914Z',
    'tcInitialized':true,
    'tc_version':1,
    'license':'CC BY-SA 4.0',
  };
  const contextId = {
    groupId: 'figs-explicit',
    occurrence: 1,
    reference: {
      'bookId': bookId,
      'chapter': 3,
      'verse': 5,
    },
  };
  const projectName = 'en_gal';
  const projectPath = path.join(PROJECTS_PATH, projectName);
  const tnIndexPath = path.join(projectPath, '.apps', 'translationCore', 'index', TRANSLATION_NOTES, bookId);

  beforeEach(() => {
    // reset mock filesystem data
    fs.__resetMockFS();
    loadMockFsWithProjectAndResources();
    const sourceResourcesPath = path.join('src', '__tests__', 'fixtures', 'resources');
    const copyResourceFiles = ['en/translationHelps/translationNotes'];
    fs.__loadFilesIntoMockFs(copyResourceFiles, sourceResourcesPath, USER_RESOURCES_PATH);
    const sourcePath = path.join('src', '__tests__/fixtures/project');
    let copyFiles = [projectName];
    fs.__loadFilesIntoMockFs(copyFiles, sourcePath, PROJECTS_PATH);
    fs.__loadFilesIntoMockFs(['source-content-updater-manifest.json'], STATIC_RESOURCES_PATH, USER_RESOURCES_PATH);
  });

  it('should succeed', () => {
    // given
    const toolName = TRANSLATION_NOTES;
    const contextId_ = _.cloneDeep(contextId);
    contextId_.tool = toolName;
    const store = mockStore({
      resourcesReducer: {
        bibles: { targetLanguage: { targetBible: { manifest: {} } } },
        translationHelps: {},
        lexicons: {},
      },
      contextIdReducer: { contextId: {} },
      settingsReducer: {
        toolsSettings: {
          ScripturePane: {
            currentPaneSettings: [
              {
                bibleId: TARGET_BIBLE,
                languageId: TARGET_LANGUAGE,
              }, {
                bibleId: 'ugnt',
                languageId: ORIGINAL_LANGUAGE,
              }, {
                bibleId: 'ust',
                languageId: 'en',
              }, {
                bibleId: 'ult',
                languageId: 'en',
              },
            ],
          },
        },
      },
      projectDetailsReducer: {
        manifest: manifest_,
        projectSaveLocation: projectPath,
      },
    });
    const loadPath = getContextIdPathFromIndex(projectPath, toolName, bookId);
    fs.outputJsonSync(loadPath, contextId_);

    // when
    store.dispatch(updateGroupIndexForGl(toolName, 'en'));

    // then
    const storedContextId = fs.readJsonSync(loadPath);
    expect(storedContextId).toMatchSnapshot();
    expect(fs.existsSync(path.join(tnIndexPath, contextId.groupId + '.json'))).toBeTruthy(); // should have copied resources
  });
});

//
// helpers
//

function cleanupResources(resourceList) {
  const newResourceList = [];

  for (let resource of resourceList) {
    const resource_ = _.cloneDeep(resource); // make copy
    expect (resource_.manifest).not.toBeUndefined();
    resource_.manifest = 'manifest';
    newResourceList.push(resource_);
  }
  return newResourceList;
}

function loadMockFsWithProjectAndResources() {
  const sourcePath = path.join('src', '__tests__', 'fixtures', 'project');
  const copyFiles = ['en_gal'];
  fs.__loadFilesIntoMockFs(copyFiles, sourcePath, PROJECTS_PATH);

  const sourceResourcesPath = path.join('src', '__tests__', 'fixtures', 'resources');
  const resourcesPath = USER_RESOURCES_PATH;
  const copyResourceFiles = [
    'en/bibles/ult',
    'en/bibles/ust',
    'el-x-koine/bibles/ugnt',
    'hbo/bibles/uhb',
    'en/translationHelps/translationWords',
    'en/translationHelps/translationAcademy',
    'hi/translationHelps/translationWords'];
  fs.__loadFilesIntoMockFs(copyResourceFiles, sourceResourcesPath, resourcesPath);
}

function loadSourceContentUpdaterManifests(bundledDate, userDate, appVersion = APP_VERSION) {
  const bundledResourcesManifestPath = path.join(STATIC_RESOURCES_PATH, 'source-content-updater-manifest.json');
  fs.ensureDirSync(STATIC_RESOURCES_PATH);

  if (bundledDate) {
    fs.outputJsonSync(bundledResourcesManifestPath, { modified: bundledDate });
  }

  const resourcesManifestPath = path.join(USER_RESOURCES_PATH, 'source-content-updater-manifest.json');
  fs.ensureDirSync(USER_RESOURCES_PATH);

  if (userDate) {
    const manifest = { modified: userDate };

    if (typeof appVersion === 'string') {
      manifest[TC_VERSION] = appVersion; // add app version to resource
    }
    fs.outputJsonSync(resourcesManifestPath, manifest);
  }
}
