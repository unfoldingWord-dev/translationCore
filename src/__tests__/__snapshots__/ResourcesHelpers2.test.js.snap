// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAvailableScripturePaneSelections getAvailableScripturePaneSelections() should work without targetBible loaded into resources 1`] = `Array []`;

exports[`getAvailableScripturePaneSelections should work 1`] = `Array []`;

exports[`getAvailableScripturePaneSelections should work OT 1`] = `Array []`;

exports[`getResourcesNeededByTool() should work 1`] = `
Array [
  Object {
    "bibleId": "ult",
    "languageId": "en",
  },
  Object {
    "bibleId": "ugnt",
    "languageId": "el-x-koine",
  },
  Object {
    "bibleId": "ust",
    "languageId": "en",
  },
]
`;

exports[`updateGroupIndexForGl() should succeed 1`] = `
Object {
  "groupId": "figs-explicit",
  "occurrence": 1,
  "occurrenceNote": "Your language may require that what the people heard and whom they trusted be stated explicitly. Alternate translation: \\"because you heard the message and had faith in Jesus\\" or \\"because you listened to the message and trusted in Jesus\\" (See: [[rc://en/ta/man/translate/figs-explicit]])",
  "reference": Object {
    "bookId": "gal",
    "chapter": 3,
    "verse": 5,
  },
  "tool": "translationNotes",
}
`;
